src/OrchardCore.Modules/Orchard.Autoroute/Handlers/AutoroutePartHandler.cs

```diff
@@ -127,26 +132,29 @@ private string GetPattern(AutoroutePart part)
 
         private string GenerateUniquePath(string path, AutoroutePart context)
         {
-            var similarPaths = _session.QueryIndexAsync<AutoroutePartIndex>(o => o.ContentItemId != context.ContentItem.ContentItemId && o.Path.StartsWith(path)).List().GetAwaiter().GetResult();
-            if (!similarPaths.Any(o => o.Path == path))
+            var version = 1;
+            var unversionedPath = path;
+
+            var versionSeparatorPosition = path.LastIndexOf('-');
+            if (versionSeparatorPosition > -1)
             {
-                return path;
+                int.TryParse(path.Substring(versionSeparatorPosition).TrimStart('-'), out version);
+                unversionedPath = path.Substring(0, versionSeparatorPosition);
             }
 
-            var version = similarPaths.Select(s => GetSlugVersion(path, s.Path)).OrderBy(i => i).LastOrDefault();
-            return version != null ? $"{path}-{version}" : path;
-        }
-
-        private int? GetSlugVersion(string path, string potentiallyConflictingPath)
-        {
-            var slugParts = potentiallyConflictingPath.Split(new[] { path }, StringSplitOptions.RemoveEmptyEntries);
-
-            if (slugParts.Length == 0)
+            while (true)
             {
-                return 1;
+                var versionedPath = $"{unversionedPath}-{version++}";
+                if (IsPathUnique(versionedPath, context))
+                {
+                    return versionedPath;
+                }
             }
+        }
 
-            return int.TryParse(slugParts[0].TrimStart('-'), out int v) ? (int?)++v : null;
+        private bool IsPathUnique(string path, AutoroutePart context)
```

Comment:
We own the full call stack on this method up to the driver, so we should make this one Async all the way up.